lib/Sema/CSSimplify.cpp

```diff
@@ -1410,6 +1410,22 @@ ConstraintSystem::matchTypes(Type type1, Type type2, ConstraintKind kind,
     case ConstraintKind::ArgumentConversion:
     case ConstraintKind::OperatorArgumentTupleConversion:
     case ConstraintKind::OperatorArgumentConversion:
+      if (type1->is<InOutType>() && typeVar2 &&
+          typeVar2->getImpl().canBeInOut()) {
```

Comment:
When `type1` is an `InOutType`, we have two cases for `typeVar2`: either it allows binding to `inout` (in which case we do the binding, as it's doing below), or it does *not* allow binding to `inout` (in which case we should immediately fail). The immediate-failure case might not be strictly necessary, but it would be nice to early-exit with the failure here rather than go through the entirety of `matchTypes`.

Also, could this happen the other way, where `type2` is an `InOutType` and `typeVar1` needs to be checked for `canBeInOut`?